Add support for `pre-wrap` and `pre-line` values for `white-space`.

eefriedman opened the pull request:
This is mostly straightforward.  I had to modify a couple of places
which were accidentally discarding whitespace.

Fixes #1513.

This fixes some relevant tests from the CSS testsuite... but a lot of
them are either manual, or don't pass because of unrelated issues.  (For
example, white-space-mixed-002 renders correctly, but
white-space-mixed-002-ref doesn't because of a float bug.)

I'd appreciate any suggestions for how to go about adding tests for this.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/7951)

<!-- Reviewable:end -->

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify layout code, but no reftests are modified. Please consider adding a reftest!

pcwalton commented on components/gfx/text/util.rs:
```diff
@@ -38,7 +38,7 @@ pub fn transform_text(text: &str,
                 }
             }
-            text.len() > 0 && is_in_whitespace(text.char_at_reverse(0), mode)
+            false
```
Why this change?

eefriedman replied:
This fixes https://github.com/servo/servo/blob/master/tests/wpt/css-tests/css21_dev/html4/white-space-mixed-002.htm and some related issues.  The return value here basically means "should leading whitespace be discarded from the next fragment", and the answer is always no in `white-space: pre` or `white-space: pre-wrap` mode.

pcwalton commented on components/layout/fragment.rs:
```diff
@@ -1275,10 +1275,18 @@ impl Fragment {
+            white_space::T::normal |
+            white_space::T::pre_wrap |
+            white_space::T::pre_line => true,
+        }
```
Matches very similar to this are duplicated throughout the code. Maybe we should have a `whitespace_is_pre_or_nowrap()` kind of function?

pcwalton replied:
Actually, never mind, I see the matches are subtly different each time.

eefriedman replied:
Hmm... there are basically three useful ways to divide up the values of white-space: "are spaces preserved", "are newlines preserved", and "is wrapping allowed".  I could add helpers for all three, which might be useful for readability even if they don't have very many uses.

pcwalton replied:
Yeah, I agree those would be useful distinctions for readability.

pcwalton commented on components/layout/fragment.rs:
```diff
@@ -1560,18 +1570,9 @@ impl Fragment {
-                pieces_processed_count += 1;
+                debug!("calculate_split_position_using_breaking_strategy: enlarging span");
+                remaining_inline_size = remaining_inline_size - advance;
+                inline_start_range.extend_by(slice.range.length());
```
Why unconditionally set `remaining_inline_size` now?

eefriedman replied:
Like the debug! log said, the first branch would trim leading whitespace.  That doesn't really work out very well in pre-wrap mode, where whitespace is supposed to be preserved.

I was going to add the condition that the fragment isn't using a whitespace-preserving mode... but then I realized we never actually end up in that situation: we strip leading whitespace before we reach this code.

pcwalton commented:
Looks good other than these couple of questions I had.

Ms2ger commented:
This needs at least some automated testing. If you know which csswg-test tests this causes to pass, feel free to submit references upstream and ask me to review. We can land this PR once I've integrated those references.

eefriedman commented:
Refactored the code a bit, and added some tests.

pcwalton commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 9202ebc has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton 9202ebc6375fcfdc1e462e45838c3f173aa6daff -->

pcwalton commented:
Looks great, thanks!

bors-servo commented:
:hourglass: Testing commit 9202ebc with merge 477440c...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/481)

eefriedman commented:
Looks like a legitimate bug in my patch.

eefriedman committed:
Add support for `pre-wrap` and `pre-line` values for `white-space`.

This is mostly straightforward.  I had to modify a couple of places
which were accidentally discarding whitespace.

Fixes #1513.

eefriedman commented:
Added a fix for newlines in `white-space: pre` mode.  (See https://github.com/servo/servo/pull/7951/files#diff-013814e2f277c3fe9e1b72e33bd8ad2aL621)  Patch should be good to go now.

eefriedman commented:
@pcwalton Ping.  This is ready to merge.

eefriedman commented:
Rebased.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #8088) made this pull request unmergeable. Please resolve the merge conflicts.

eefriedman commented:
Rebased again.

@pcwalton ping.

pcwalton commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 3a451ff has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton 3a451ff84517f63faaf2d3439ee56748f162eb9f -->

bors-servo commented:
:hourglass: Testing commit 3a451ff with merge c3ab711...

pcwalton commented:
Sorry this fell through the cracks!

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/3500), [gonk](http://build.servo.org/builders/gonk/builds/2661), [linux-dev](http://build.servo.org/builders/linux-dev/builds/580), [linux-rel](http://build.servo.org/builders/linux-rel/builds/602), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/616), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/599), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/635)
